modules/API/API.cpp

```diff
@@ -0,0 +1,191 @@
+
+        // API variables
+        std::string var_base = evse_base + "/var/";
+
+        std::string var_powermeter = var_base + "powermeter";
+        evse->subscribe_powermeter([this, var_powermeter, &session_info](Object powermeter) {
+            json powermeter_json = powermeter;
+            this->mqtt.publish(var_powermeter, powermeter_json.dump());
+            session_info->set_latest_energy_wh(powermeter_json.at("energy_Wh_import").at("total"));
+            session_info->set_latest_total_w(powermeter_json.at("power_W").at("total"));
+        });
+
+        std::string var_datetime = var_base + "datetime";
+        std::string var_session_info = var_base + "session_info";
+        this->datetime_thread = std::thread([this, var_datetime, var_session_info, &session_info]() {
+            auto next_tick = std::chrono::system_clock::now();
+            while (this->running) {
+                std::string datetime_str = to_rfc3339(std::chrono::system_clock::now());
+                this->mqtt.publish(var_datetime, datetime_str);
+                this->mqtt.publish(var_session_info, *session_info);
+
+                next_tick += std::chrono::seconds(1);
+                std::this_thread::sleep_until(next_tick);
+            }
+        });
+
+        std::string var_events = var_base + "events";
+        evse->subscribe_session_events([this, var_events, var_session_info, &session_info](Object session_events) {
+            auto state = this->sanitize_event(session_events["event"].get<std::string>());
+            this->mqtt.publish(var_events, state);
+            session_info->set_state(state);
+            if (state == "SessionStarted") {
+                auto session_started = session_events["session_started"];
+                auto energy_Wh_import = session_started["energy_Wh_import"].get<double>();
+                session_info->set_start_energy_wh(energy_Wh_import);
+            } else if (state == "SessionFinished") {
+                auto session_finished = session_events["session_finished"];
+                auto energy_Wh_import = session_finished["energy_Wh_import"].get<double>();
+                session_info->set_end_energy_wh(energy_Wh_import);
+            }
+
+            this->mqtt.publish(var_session_info, *session_info);
+        });
+
+        // API commands
+        std::string cmd_base = evse_base + "/cmd/";
+
+        std::string cmd_pause_charging = cmd_base + "pause_charging";
+        this->mqtt.subscribe(cmd_pause_charging, [&evse](const std::string& data) {
+            evse->call_pause_charging(); //
+        });
+
+        std::string cmd_resume_charging = cmd_base + "resume_charging";
+        this->mqtt.subscribe(cmd_resume_charging, [&evse](const std::string& data) {
+            evse->call_resume_charging(); //
+        });
+
+        count += 1;
+    }
+}
+
+void API::ready() {
+    invoke_ready(*p_main);
+}
+
+std::string API::sanitize_event(const std::string& event) {
```

Comment:
From my understanding this function converts the time series of events that comes in from EvseManager into an internal state of this module that is useful for display, so I would suggest to rename this function to e.g. update_state(event) or something better. I think it is important as the difference between events and states can be confusing anyway. This function may actually update several internal states in the future such as reservation_state as well as the events to control this one also arrive here.

Reply:
Sounds good